system-helper/flatpak-system-helper.c

```diff
@@ -1151,6 +1291,343 @@ handle_run_triggers (FlatpakSystemHelper   *object,
 }
 
+static gboolean
+get_connection_uid (GDBusMethodInvocation *invocation, uid_t *out_uid, GError **error)
+{
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  const gchar *sender = g_dbus_method_invocation_get_sender (invocation);
+  g_autoptr(GVariant) dict = NULL;
+  g_autoptr(GVariant) credentials = NULL;
+
+  credentials = g_dbus_connection_call_sync (connection,
+                                             "org.freedesktop.DBus",
+                                             "/org/freedesktop/DBus",
+                                             "org.freedesktop.DBus",
+                                             "GetConnectionCredentials",
+                                             g_variant_new ("(s)", sender),
+                                             G_VARIANT_TYPE ("(a{sv})"), G_DBUS_CALL_FLAGS_NONE,
+                                             G_MAXINT, NULL, error);
+  if (credentials == NULL)
+    return FALSE;
+
+  dict = g_variant_get_child_value (credentials, 0);
+
+   if (!g_variant_lookup (dict, "UnixUserID", "u", out_uid))
+    {
+      g_set_error (error, G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                   "Failed to query UnixUserID for the bus name: %s", sender);
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+check_for_system_helper_user (struct passwd  *passwd,
+                              gchar         **passwd_buf,
+                              GError        **error)
+{
+  struct passwd *result = NULL;
+  g_autofree gchar *buf = NULL;
+  size_t bufsize;
+  int err;
+
+  bufsize = sysconf (_SC_GETPW_R_SIZE_MAX);
+  if (bufsize == -1)          /* Value was indeterminate */
+     bufsize = 16384;        /* Should be more than enough */
+
+  while (!result)
+    {
+      buf = g_malloc0 (bufsize);
+      err = getpwnam_r (SYSTEM_HELPER_USER, passwd, buf, bufsize, &result);
+      if (result == NULL)
+        {
+          if (err == ERANGE)     /* Insufficient buffer space */
+            {
+              g_free (buf);
+              bufsize *= 2;
+              continue;
+            }
+          else if (err == 0)     /* User SYSTEM_HELPER_USER 's record was not found*/
+            {
+              g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND,
+                           "User %s does not exist in password file entry", SYSTEM_HELPER_USER);
+              return FALSE;
+            }
+          else
+            {
+              g_set_error (error, G_IO_ERROR, g_io_error_from_errno (err),
+                           "Failed to query user %s from password file entry", SYSTEM_HELPER_USER);
+              return FALSE;
+            }
+        }
+    }
+
+  *passwd_buf = g_steal_pointer (&buf);
+
+  return TRUE;
+}
+
+static void
+revokefs_fuse_backend_child_setup (gpointer user_data)
+{
+  struct passwd *passwd = user_data;
+
+  if (setgid (passwd->pw_gid) == -1)
+    g_warning ("Failed to setgid(%d) for revokefs backend: %s",
```

Comment:
I think we should fail harder here. If we can't setuid or setgid, then we're still root and going, which is risky. Better to just abort the process and fail some safer way.

Reply:
hmm, I agree with the point. Although, I am not sure what should be the correct implementation.

In the latest fixup, I added a `exit (1)` on the error path. I tested briefly to trigger it, and it avoids spawning the full-fleged backend. `CancelPull` is triggered too so cleanup is done and pull fallbacks to the current (temporary-double-disk) code path.